### ISSUE TITLE
JSP compilation may result in re-compilation of embedded source files for JSP-referenced classes

### ISSUE THREAD

**Issue by gregn123** [javaee/glassfish#22211]
## Environment Details
* GlassFish Version (and build number): Glassfish5 (and Glassfish 4.x), all builds
* JDK version: JDK8 (and JDK7 for GF4.x)
* OS: Windows (e.g. Windows 7, Server 2012), Linux
* Database: not relevant

----------

## Problem Description
It is common practice, at least during development, to include java sources in an application, for debugging purposes for example.
I have found that if such source files are included as part of a JSP-based JavaEE application, and the corresponding classes are referenced or invoked by the JSP, then those sources may be re-compiled as part of the compilation of the (servlet) source file generated by the JSP compiler. This compilation of other sources files shouldn't be happening - and not to mention, those sources have ALREADY been compiled and their .class files included in the application. Only the servlet class source code, generated from the JSP by the JSP compiler, is meant to be compiled.

The problem basically boils down to the following factors:

1) Glassfish doesn't maintain the "lastModified" time of the files that it unpacks to the filesystem from application archives (.JAR, .EAR, .WAR etc.) on deployment. The last-modification time is simply the extraction time. So the timestamps on the files depend on the order of extraction, and this may vary (e.g. xxxx.java may be extracted after xxxx.class, even though xxxx.java has an earlier timestamp in the JAR file than xxxx.class). Also, since extraction happens so quickly, many files extracted to the same directory have exactly the SAME timestamp.

2) When the JSP compiler invokes the JavaC API to compile the .java file (xxxx_jsp.java) that it has generated from xxxx.jsp, it will also compile source for an implicitly referenced class (e.g. an EJB persistence class invoked by the JSP), if the .java source file has a last-modified time more recent than the corresponding .class file. If the last-modified times of the .java and .class files are the same, then the .java file MAY be compiled, depending upon the "order of discovery" of the .java and .class files in the parsing process that occurs as part of compiling the generated servlet code (xxxx_jsp.java). The java compiler depends on timestamps in determining whether implicitly-referenced classes should have their sources (if supplied) compiled. Unfortunately there seems to be no javac option to always use the .class file in preference to the .java file, if both are present. The "-Xprefer" option only accepts "source" or "newer" as a parameter ("-Xprefer:class" is not supported).

3) Embedded sources may not actually be re-compilable at runtime, because the compilation classpath used by the JSP compiler may not include all the necessary classes, that were present when the .class files were originally built.

Because of these factors, it is possible to have, for example, a JSP-based application with embedded EJB sources which "sometimes works and sometimes doesn't". That is, sometimes the JSP compiles OK and the application works fine, and sometimes the JSP fails to compile (i.e. because an attempt is made to compile source included for a dependent class, and it fails to re-compile that source code) and the application fails. This presents itself as a "random JSP compilation failure".

## Steps to reproduce
I have attached a sample application which demonstrates this problem (customer-cmp-ear.ear). The sample application was developed by Sun Microsystems and is from an old set of Glassfish JavaEE5 samples. I updated the EAR file to include the corresponding source files for the classes in the Ejb JAR file (customer-cmp-ejb.jar) within the EAR file.

The application source code can be found here (see the "ws/javaee5/enterprise/customer-cmp-ear" directory):

[https://github.com/javaee/glassfish-samples](https://github.com/javaee/glassfish-samples)

In order to detect compilation of the EJB's "Customer.java" persistence class source code, which may erroneously occur as a result of JSP compilation, I've updated the NamedQueries annotation in that source code to include a default QueryHint that uses an EclipseLink class that isn't in the classpath of the JSP compiler, and so will catch out any attempted compilation of that source code, by resulting in a compilation error (see below, for the "hints=..." addition to the annotation).

```
@NamedQueries(
    value={@NamedQuery(name="findCustomerByFirstName", query="select object(c) from Customer c where c.firstName= :firstName",hints={@QueryHint(name=org.eclipse.persistence.config.QueryHints.PESSIMISTIC_LOCK, value=org.eclipse.persistence.config.PessimisticLock.NoLock)}),
    @NamedQuery(name="findCustomerByLastName", query="select object(c) from Customer c where c.lastName= :lastName")}
)
```

Without adding this (to cause a compilation error if it gets compiled) we'd otherwise have to check the modification time of the Customer.class file in the "glassfish\domains\domain1\applications\..." directory, to see if we can detect an update to the .class file (as a result of the source code getting compiled) after the time it was unpacked on deployment.

The following is the list of files I've attached (contained in [jsp_compile_error.zip](https://github.com/javaee/glassfish/files/1223871/jsp_compile_error.zip)), to be used for setting up and reproducing the problem:

| File | Purpose |
| ---- | --------- |
| customer-cmp-ear.ear       | the test application |
| startall.bat                          | Windows batch file to start the server and database
| setup.bat                            | Windows batch file to create necessary resources and deploy the application
| redeploy.bat                       | Windows batch file to redeploy the application (so the next JSP access will result in JSP compilation)
| remove.bat                         | Windows batch file to undeploy the test application and delete associated resources and tables
| stopall.bat                          | Windows batch file to stop the server and database
| jsp_compile_fails.jpg          | screen capture when the JSP fails to compile
| jsp_compile_succeeds.jpg  | screen capture when the JSP compiles OK
| ear_with_no_source            | directory containing the version of customer-cmp-ear.ear with no source code included
| ear_with_source                  | directory containing the version of customer-cmp-ear.ear with source code included
| linux                                    | directory containing equivalent scripts for Linux (for testing on Linux, if desired ...)
| test_jsp.bat                         | Windows batch file to test accessing the createCustomer JSP in a browser (Internet Explorer) - tests 10 accesses
| launch_url.vbs                    | Windows VBScript for launching the JSP URL in Internet Explorer

Follow the steps below to setup and run the test application (assumes Glassfish is installed):

1)	Ensure that JDK8 is set in your PATH environment variable. 
2)	Set the GF_HOME environment variable to the Glassfish installation directory (e.g. C:\glassfish5). 
3)	Run the "startall" script/batch-file to start the server and (JavaDB) database.
4)	Run the "setup" script/batch-file to create the necessary JDBC resources and deploy the application.
5)	Open http://localhost:8080/customer-cmp-war/createCustomer.jsp in a web-browser.
6)	If the JSP compiles OK, the page will be displayed (see "jsp_compile_succeeds.jpg"). If the JSP fails to compile, errors will be displayed (see "jsp_compile_fails.jpg").
7)	To re-test, you need to first redeploy the application by running the "redeploy" script/batch-file and then refresh the web page (you cannot just refresh the page, as the server will not attempt to recompile the JSP again, unless it is redeployed).
8)	Finally (after you've finished testing), to undeploy, delete resources and stop everything, run the "remove" and then the "stopall" scripts/batch-files.

It may take several redeploys and web-page accesses in order to see the compilation error occur (indicating compilation of source code that shouldn't be getting compiled).
For some reason, the problem seemed to reproduce much more easily under Windows 2012 Server (perhaps because I had it installed in a VM and it ran a bit slower than a native environment).
Under Windows 7, I usually saw the error within about 20 redeploys and JSP-page accesses. Any load on the system or other system activity seemed to help reproduce the problem, as it slows down the deployment and unpacking of files to the filesystem.

Because of this, instead of manually running steps 5 - 7 above, you might want to just run "test_jsp.bat", which redeploys and launches the JSP URL in Internet Explorer, 10 times consecutively (note: creates 10 IE windows, which you'll need to close). This can easily be adjusted if required.

When the problem occurs, you see an HTTP 500 error and compilation error details reported in the browser window, like the one below:

```
HTTP Status 500 - Internal Server Error


type Exception report

message Internal Server Error

description The server encountered an internal error that prevented it from fulfilling this request.

exception 
org.apache.jasper.JasperException: PWC6033: Error in Javac compilation for JSP

PWC6199: Generated servlet error:
source value 1.5 is obsolete and will be removed in a future release

PWC6199: Generated servlet error:
target value 1.5 is obsolete and will be removed in a future release

PWC6199: Generated servlet error:
To suppress warnings about obsolete options, use -Xlint:-options.

PWC6197: An error occurred at line: 44 in the jsp file: /createCustomer.jsp
PWC6199: Generated servlet error:
package org.eclipse.persistence.config does not exist

PWC6197: An error occurred at line: 44 in the jsp file: /createCustomer.jsp
PWC6199: Generated servlet error:
package org.eclipse.persistence.config does not exist

PWC6199: Generated servlet error:
C:\glassfish5\glassfish\domains\domain1\applications\cmp-jpa\customer-cmp-ejb_jar\enterprise\customer_cmp_ejb\persistence\Customer.java uses unchecked or unsafe operations.

PWC6199: Generated servlet error:
Recompile with -Xlint:unchecked for details.
```

In the server.log the following stacktrace can be seen:

```
[2017-08-14T17:40:23.951+1000] [glassfish 5.0] [SEVERE] [] [org.apache.jasper.compiler.Compiler] [tid: _ThreadID=30 _ThreadName=http-listener-1(1)] [timeMillis: 1502696423951] [levelValue: 1000] [[
  Error compiling file: C:\glassfish5\glassfish\domains\domain1\generated\jsp\cmp-jpa\customer-cmp-war_war\org\apache\jsp\createCustomer_jsp.java]]

[2017-08-14T17:40:23.952+1000] [glassfish 5.0] [WARNING] [] [javax.enterprise.web] [tid: _ThreadID=30 _ThreadName=http-listener-1(1)] [timeMillis: 1502696423952] [levelValue: 900] [[
  StandardWrapperValve[jsp]: Servlet.service() for servlet jsp threw exception
org.apache.jasper.JasperException: PWC6033: Error in Javac compilation for JSP

PWC6199: Generated servlet error:
source value 1.5 is obsolete and will be removed in a future release

PWC6199: Generated servlet error:
target value 1.5 is obsolete and will be removed in a future release

PWC6199: Generated servlet error:
To suppress warnings about obsolete options, use -Xlint:-options.

PWC6197: An error occurred at line: 44 in the jsp file: /createCustomer.jsp
PWC6199: Generated servlet error:
package org.eclipse.persistence.config does not exist

PWC6197: An error occurred at line: 44 in the jsp file: /createCustomer.jsp
PWC6199: Generated servlet error:
package org.eclipse.persistence.config does not exist

PWC6199: Generated servlet error:
C:\glassfish5\glassfish\domains\domain1\applications\cmp-jpa\customer-cmp-ejb_jar\enterprise\customer_cmp_ejb\persistence\Customer.java uses unchecked or unsafe operations.

PWC6199: Generated servlet error:
Recompile with -Xlint:unchecked for details.


	at org.apache.jasper.compiler.DefaultErrorHandler.javacError(DefaultErrorHandler.java:129)
	at org.apache.jasper.compiler.ErrorDispatcher.javacError(ErrorDispatcher.java:299)
	at org.apache.jasper.compiler.Compiler.generateClass(Compiler.java:392)
	at org.apache.jasper.compiler.Compiler.compile(Compiler.java:453)
	at org.apache.jasper.JspCompilationContext.compile(JspCompilationContext.java:625)
	at org.apache.jasper.servlet.JspServletWrapper.service(JspServletWrapper.java:375)
	at org.apache.jasper.servlet.JspServlet.serviceJspFile(JspServlet.java:473)
	at org.apache.jasper.servlet.JspServlet.service(JspServlet.java:377)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:791)
	at org.apache.catalina.core.StandardWrapper.service(StandardWrapper.java:1580)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:258)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:160)
	at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:652)
	at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:591)
	at com.sun.enterprise.web.WebPipeline.invoke(WebPipeline.java:99)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:155)
	at org.apache.catalina.connector.CoyoteAdapter.doService(CoyoteAdapter.java:371)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:238)
	at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:460)
	at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:168)
	at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:206)
	at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:180)
	at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:242)
	at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:119)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:284)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:201)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:133)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:112)
	at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:77)
	at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:539)
	at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:112)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:117)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:56)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:137)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:593)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:573)
	at java.lang.Thread.run(Thread.java:748)
]]
```

Despite the error message claiming that "Error compiling file: ... createCustomer_jsp.java", the error is not from compiling that source file, it's actually from an attempt to compile the EJB-related source code file "Customer.java", which fails to compile because "package org.eclipse.persistence.config does not exist". Of course, JSP compilation should NOT result in this source code file being (re)compiled, it's only meant to be compiling the source code (servlet class) generated from the JSP.

When the EJB source code isn't included in the EAR file (i.e. if you use "customer-cmp-ear.jar" from the "ear_with_no_source" directory), then the JSP compilation error never occurs when the JSP is compiled, and the application always works fine.

## Impact of Issue
A "random" JSP compilation error occurs and the application terminates with a HTTP 500 error. Sometimes it occurs frequently, sometimes about 10% of the time. Of course, I can just remove source code from the application, and the problem won't occur. But the point is that it is possible for the JSP compiler to trigger compilation of source code that it shouldn't be compiling. The compilation will silently go unnoticed if that source can be successfully compiled using the JSP's compiler's classpath. There are also performance implications, and possibly even security-related implications - which I have not further explored.

**Comment by ruolli** [comment]
@gregn123  I put some effort on this issue in the past days, and would like to paste my findings here for further discussion. 
1) The recompilation behavior is the JavaCompiler default behavior. 
```
    /** Include class corresponding to given class file in package,
     *  unless (1) we already have one the same kind (.class or .java), or
     *         (2) we have one of the other kind, and the given class file
     *             is older.
     */
    protected void includeClassFile(PackageSymbol p, JavaFileObject file) {
        if ((p.flags_field & EXISTS) == 0)
            for (Symbol q = p; q != null && q.kind == PCK; q = q.owner)
                q.flags_field |= EXISTS;
        JavaFileObject.Kind kind = file.getKind();
        int seen;
        if (kind == JavaFileObject.Kind.CLASS)
            seen = CLASS_SEEN;
        else
            seen = SOURCE_SEEN;
        String binaryName = fileManager.inferBinaryName(currentLoc, file);
        int lastDot = binaryName.lastIndexOf(".");
        Name classname = names.fromString(binaryName.substring(lastDot + 1));
        boolean isPkgInfo = classname == names.package_info;
        ClassSymbol c = isPkgInfo
            ? p.package_info
            : (ClassSymbol) p.members_field.lookup(classname).sym;
        if (c == null) {
            c = enterClass(classname, p);
            if (c.classfile == null) // only update the file if's it's newly created
                c.classfile = file;
            if (isPkgInfo) {
                p.package_info = c;
            } else {
                if (c.owner == p)  // it might be an inner class
                    p.members_field.enter(c);
            }
        } else if (c.classfile != null && (c.flags_field & seen) == 0) {
            // if c.classfile == null, we are currently compiling this class
            // and no further action is necessary.
            // if (c.flags_field & seen) != 0, we have already encountered
            // a file of the same kind; again no further action is necessary.
            if ((c.flags_field & (CLASS_SEEN | SOURCE_SEEN)) != 0)
                c.classfile = preferredFileObject(file, c.classfile);
        }
        c.flags_field |= seen;
    }

    /** Implement policy to choose to derive information from a source
     *  file or a class file when both are present.  May be overridden
     *  by subclasses.
     */
    protected JavaFileObject preferredFileObject(JavaFileObject a,
                                           JavaFileObject b) {

        if (preferSource)
            return (a.getKind() == JavaFileObject.Kind.SOURCE) ? a : b;
        else {
            long adate = a.getLastModified();
            long bdate = b.getLastModified();
            // 6449326: policy for bad lastModifiedTime in ClassReader
            //assert adate >= 0 && bdate >= 0;
            return (adate > bdate) ? a : b;
        }
    }

```

**Comment by ruolli** [comment]
I compared the behavior of WebLogic. WebLogic uses ECJ (Eclipse Compiler for Java), not JDK JavaCompiler. 
The default behavior of ECJ and JDK JavaCompiler has no difference, i.e. , both the compilers would compare the last modified time of *.java and *.class before loading any of it. If the *.java is newer, the compilers would compile the *.java to overwrite existing *.class before loading the *.class .
`com.sun.tools.javac.jvm.ClassReader` snippet : 

```
    /** Include class corresponding to given class file in package,
     *  unless (1) we already have one the same kind (.class or .java), or
     *         (2) we have one of the other kind, and the given class file
     *             is older.
     */
    protected void includeClassFile(PackageSymbol p, JavaFileObject file) {
        if ((p.flags_field & EXISTS) == 0)
            for (Symbol q = p; q != null && q.kind == PCK; q = q.owner)
                q.flags_field |= EXISTS;
        JavaFileObject.Kind kind = file.getKind();
        int seen;
        if (kind == JavaFileObject.Kind.CLASS)
            seen = CLASS_SEEN;
        else
            seen = SOURCE_SEEN;
        String binaryName = fileManager.inferBinaryName(currentLoc, file);
        int lastDot = binaryName.lastIndexOf(".");
        Name classname = names.fromString(binaryName.substring(lastDot + 1));
        boolean isPkgInfo = classname == names.package_info;
        ClassSymbol c = isPkgInfo
            ? p.package_info
            : (ClassSymbol) p.members_field.lookup(classname).sym;
        if (c == null) {
            c = enterClass(classname, p);
            if (c.classfile == null) // only update the file if's it's newly created
                c.classfile = file;
            if (isPkgInfo) {
                p.package_info = c;
            } else {
                if (c.owner == p)  // it might be an inner class
                    p.members_field.enter(c);
            }
        } else if (c.classfile != null && (c.flags_field & seen) == 0) {
            // if c.classfile == null, we are currently compiling this class
            // and no further action is necessary.
            // if (c.flags_field & seen) != 0, we have already encountered
            // a file of the same kind; again no further action is necessary.
            if ((c.flags_field & (CLASS_SEEN | SOURCE_SEEN)) != 0)
                c.classfile = preferredFileObject(file, c.classfile);
        }
        c.flags_field |= seen;
    }

    /** Implement policy to choose to derive information from a source
     *  file or a class file when both are present.  May be overridden
     *  by subclasses.
     */
    protected JavaFileObject preferredFileObject(JavaFileObject a,
                                           JavaFileObject b) {

        if (preferSource)
            return (a.getKind() == JavaFileObject.Kind.SOURCE) ? a : b;
        else {
            long adate = a.getLastModified();
            long bdate = b.getLastModified();
            // 6449326: policy for bad lastModifiedTime in ClassReader
            //assert adate >= 0 && bdate >= 0;
            return (adate > bdate) ? a : b;
        }
    }
```

**Comment by ruolli** [comment]
A further comparison between JDK Compiler and ECJ. 
JDK Compiler has the `preferSource` variable to prioritize *.java over *.class , while ECJ has `asBinaryOnly` variable to prioritize *.class over *.java . Neither compiler has its opposite variable to prioritize reversely. 

So, to break the compilation of referenced *.java on GlassFish, it's not possible to implement simply by setting a variable, because JDK JavaCompiler does not have `perferSource` variable ...

**Comment by ruolli** [comment]
@gregn123  Would you help check whether the QueryHints is invoked at your side ? I didn't see it got invoked under the successful scenario. Thanks.

**Comment by gregn123** [comment]
I'm sorry, I don't really understand your request.
Using the QueryHints annotation was just a way to trigger a compilation error, to show that the JSP compiler was causing a Java source file to be compiled, that shouldn't be getting compiled (since that source code file had already been successfully compiled, and its .class file included in the application, for deployment). As far as I am aware, only the Java source code file generated from the JSP file, should be getting compiled by the JSP compiler, and nothing else.
Equivalently (instead of using the QueryHints annotation) you could just modify that source code file so that it is not compilable (e.g. put a syntax error in it). The point is, that source code, if included in the applicaton, shouldn't be getting compiled.
So the use of the "QueryHints" annotation is not directly related to the core problem. I just happened to use QueryHints in my test application, and include corresponding source code files for debugging purposes, and it alerted me to the fact that the JSP compiler was attempting to "randomly" re-compile one of my already-compiled classes (resulting in application failure in this case).

**Comment by ruolli** [comment]
Hi, @gregn123 , 

Ok, I got your point, please ignore my previous request about QueryHints. 
Speaking of the re-compilation behavior of JSP compiler, I think we could check two aspects : 
1) why the re-compilation of referenced *.java happened ? 
   Now, I've checked the source code of JDK JavaCompiler, found that it's the default behavior of JDK JavaCompiler to compile *.java if its newer that corresponding *.class . 
   The point is, does GlassFish have any mechanism to break the re-compilation.
   As I mentioned previously, WebLogic has its way of archive referenced files into a *.jar, and load *.class from it only, when compiling JSP pages.

**Comment by gregn123** [comment]
Hi @ruolli 

1. I am not aware of anything that Glassfish can (easily) do to prevent recompilation of .java sources in this case.
There currently seems to be no java compiler options available, for Glassfish's JSP compiler to specify, to avoid processing of included .java source files of dependent classes in this case.
Perhaps any embedded source code could be stripped out when the app is deployed to the filesystem (so it is not visible to the JSP compilation process) but otherwise made available on the classpath at runtime; easy to say, but maybe not so easy to do.
Glassfish doesn't maintain the original timestamps (last modification time) of the files within archives (.jar,.war etc.) when they get unpacked to the filesystem. Even if it did, I could still cause the .java source code to be re-compiled in this case, by manually ensuring that its timestamp was later than the .class file (e.g. using "touch") prior to creating the archive.

2. I understand your thoughts about whether QueryHints was actually getting referenced at runtime; I don't know for sure whether it's actually getting referenced, but I'd expect that class to be in the runtime CLASSPATH since it's in "org.eclipse.persistence.core.jar", and EclipseLink is the persistence provider being used.
However, when the compilation fails, the EclipseLink JAR (org.eclipse.persistence.core.jar) containing the QueryHints class is definitely NOT in the compilation CLASSPATH (I have verified this by debugging). The reported error "package org.eclipse.persistence.config does not exist" also clearly indicates this.
As I've pointed out, the use of EclipseLink-specific QueryHints in the application is not directly related to the re-compilation bug I've reported here - it's use (resulting in a compilation error) just happened to help indicate that re-compilation of an embedded source file was occurring.
